### PR TITLE
Cap lender factor at 10x the interest factor for borrowers

### PR DESCRIPTION
**Pull request by mattcushman**
# Description of change
## High level
* Cap interest rate factor for lenders as 10x that of borrowers
* Revert if awarded LP is 0 for add collateral / quote token, move quote token and merge NFT collateral

# Description of bug or vulnerability and solution
* An Ajna pool can be in a state with arbitrarily large loan size yet arbitrarily small total deposit on the book.  In these cases, the interest flowing the deposits could be very large, resulting in arbitrarily large scale factors applied to the deposits in the interest accrual operation.  This creates exchange rate instabilities and can result in numerical precision issues and roundoff error.
* Tp fix this, we cap the factor applied to depositors by 10x the factor applied to borrowers.
* One particular regression test that exposed this issue is added.  In that test, run as `eval QUOTE_PRECISION=18 COLLATERAL_PRECISION=18 BUCKET_INDEX_ERC20=2570 BUCKET_INDEX_ERC721=850 NO_OF_BUCKETS=3 forge t --mt test_regression_QT1_LPs -vv`, the large interest accrual resulted in a subsequent `addCollateral` call being awarded 0 LPs.  This fixes that error. 
*  Revert with `InsufficientLP` error if awarded LP on following lender actions is calculated as 0 (due to rounding issue)
   * add collateral
   * add quote token
   * move quote token (LP in to bucket)
   * mergeOrRemoveCollateral
  
# Contract size
## Pre Change
```
LenderActions            -   9,851B  (40.08%)
PoolCommons              -   9,382B  (38.17%)
```
## Post Change
```
LenderActions            -   9,995B  (40.67%)
PoolCommons              -   9,426B  (38.35%)
```

# Gas usage
```
| Function Name                        | min             | avg    | median | max     | # calls |
| addQuoteToken                        | 119799          | 176906 | 161624 | 838067  | 60004   |
| bucketTake                           | 311436          | 329814 | 329814 | 348192  | 4       |
| drawDebt                             | 247478          | 292132 | 266312 | 801936  | 136003  |
| kick                                 | 151711          | 221290 | 219662 | 1031014 | 48000   |
| kickWithDeposit                      | 195774          | 273194 | 269338 | 995593  | 8000    |
| moveQuoteToken                       | 275972          | 275972 | 275972 | 275972  | 1       |
| removeQuoteToken                     | 142385          | 162645 | 168988 | 185488  | 4000    |
| repayDebt                            | 573569          | 605711 | 595607 | 753878  | 16002   |
| settle                               | 347924          | 347924 | 347924 | 347924  | 1       |
| take                                 | 80551           | 81966  | 81656  | 312546  | 15998   |

ERC20Pool
| addCollateral                        | 716             | 163749 | 144791 | 397709 | 29      |
| addQuoteToken                        | 714             | 269103 | 189631 | 710408 | 680     |
| moveQuoteToken                       | 692             | 136608 | 119609 | 629412 | 38      |

ERC721Pool
| addCollateral                          | 22828           | 230310 | 169039 | 733939   | 8       |
| addQuoteToken                          | 37654           | 275452 | 194867 | 677326   | 187     |
| mergeOrRemoveCollateral                | 5867            | 120605 | 160404 | 224187   | 15      |
| moveQuoteToken                         | 40064           | 45004  | 40232  | 64374    | 5       |
```
## Post Change
```
| Function Name                        | min             | avg    | median | max     | # calls |
| addQuoteToken                        | 119822          | 176963 | 161647 | 711343  | 60004   |
| bucketTake                           | 311870          | 330248 | 330248 | 348626  | 4       |
| drawDebt                             | 247478          | 292164 | 266312 | 802479  | 136003  |
| kick                                 | 151711          | 221290 | 219662 | 1031557 | 48000   |
| kickWithDeposit                      | 195774          | 273194 | 269338 | 996136  | 8000    |
| moveQuoteToken                       | 267308          | 267308 | 267308 | 267308  | 1       |
| removeQuoteToken                     | 142820          | 163135 | 167054 | 188416  | 4000    |
| repayDebt                            | 574112          | 606254 | 596150 | 754421  | 16002   |
| settle                               | 348359          | 348359 | 348359 | 348359  | 1       |
| take                                 | 80551           | 81966  | 81656  | 312980  | 15998   |

ERC20Pool
| addCollateral                        | 716             | 163837 | 145744 | 398278 | 29      |
| addQuoteToken                        | 714             | 262833 | 189654 | 710974 | 701     |
| moveQuoteToken                       | 692             | 146284 | 121313 | 629978 | 38      |

ERC721Pool
| addCollateral                          | 22828           | 230391 | 169052 | 734508  | 8       |
| addQuoteToken                          | 37654           | 275486 | 194890 | 677892  | 187     |
| mergeOrRemoveCollateral                | 5867            | 120605 | 160404 | 224187  | 15      |
| moveQuoteToken                         | 40064           | 45007  | 40232  | 64392   | 5       |
```

### CODE REVIEW

File: src/libraries/external/PoolCommons.sol [hunk 1]
```diff
@@ -248,10 +248,13 @@ library PoolCommons {
             );
 
             // Scale the fenwick tree to update amount of debt owed to lenders
+            // lender factor computation, capped at 10x the interest factor for borrowers
+            uint256 lenderFactor = Maths.min(Maths.floorWdiv(newInterest_, interestEarningDeposit) + Maths.WAD,
+                                             Maths.wmul(pendingFactor, 10e18));
```
Comment (grandizzy) [inline src/libraries/external/PoolCommons.sol]: for clarity let's compare with `Maths.wmul(pendingFactor, 10 * Maths.WAD))` as in line above

**Review by grandizzy — APPROVED** [review]
LGTM. gas usage differs (up and down comparing with develop) but I assume that's because of the way test scenarios are done (some cases will cap to 10x so `Deposits.mult` won't consume same amount of gas)

**Comment by grandizzy** [comment]
also, unrelated but worth flagging
```
forge test -vv --match-test testLoadFenwickFuzzyScalingFind

Running 1 test for tests/forge/unit/FenwickTree.t.sol:FenwickTreeTest
[FAIL. Reason: Assertion failed. Counterexample: calldata=0x6f82d454000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff0000000000000000000000000000000000000000000000000000000000000000, args=[1766847064778384329583297500742918515827483896875618958121606201292619776, 0, 0, 115792089237316195423570985008687907853269984665640564039457584007913129639935, 0]] testLoadFenwickFuzzyScalingFind(uint256,uint256,uint256,uint256,uint256) (runs: 286, μ: 62834407, ~: 61426888)
Logs:
  Bound Result 1002
  Bound Result 1000000000000000000
  Bound Result 0
  Bound Result 7388
  Bound Result 1000000000000000000
  Error: a <= b not satisfied [uint]
    Value a: 200937085584
    Value b: 2

Test result: FAILED. 0 passed; 1 failed; finished in 35.62s
```

**Comment by grandizzy** [comment]
converting to draft in order to include reverts if addCollateral / removeCollateral awards 0 LPs

File: src/libraries/external/LenderActions.sol [hunk 2]
```diff
@@ -536,6 +548,7 @@ library LenderActions {
      *  @dev      increment `lender.lps` accumulator and `lender.depositTime` state
      *  @dev    === Reverts on ===
      *  @dev    invalid merge index `CannotMergeToHigherPrice()`
+     *  @dev    no LP awarded in bucket `InsufficientLP()`
```
Comment (EdNoepel) [inline src/libraries/external/LenderActions.sol]: would be nice if comment explicitly cited `toIndex_` instead of ambiguous `bucket`

**Review by EdNoepel — APPROVED** [review]
Just one recommendation regarding a comment.  Approved regardless.

File: src/libraries/external/LenderActions.sol (hunk 2 above)
Comment (grandizzy) [reply src/libraries/external/LenderActions.sol]: yep, good point, updated with https://github.com/ajna-finance/contracts/pull/772/commits/127ff740df8360c5e41b931ae82498e71dfe9ea4

**Review by ith-harvey — APPROVED** [review]
Originally I thought this could have been a bug where a partial NFT could get locked in a bucket because lenders don't have LP in the bucket but this is an incorrect understanding as someone could very easily come along, deposit QT and then gain enough LP to remove that dust collateral amount.

LGTM

**Review by mattcushman — COMMENTED** [review]
lgtm

**Pull request merged**